OpenRA.Mods.Common/Traits/AutoTarget.cs

```diff
@@ -168,6 +168,13 @@ Actor ChooseTarget(Actor self, WDist range, bool allowMove)
 			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, range);
 			foreach (var actor in actorsInRange)
 			{
+				// PERF: Most units can only attack enemy units. If this is the case but the target is not an enemy, we
+				// can bail early and avoid the more expensive targeting checks and armament selection. For groups of
+				// allied units, this helps significantly reduce the cost of auto target scans. This is important as
+				// these groups will continuously rescan their allies until an enemy finally comes into range.
+				if (!actor.AppearsHostileTo(self) && attack.UnforcedAttackCanOnlyTargetEnemies())
```

Comment:
`attack.UnforcedAttackCanOnlyTargetEnemies()` can be hoisted outside the loop for even more perf.